design-proposals/export-import/volume-export.md

```diff
@@ -0,0 +1,161 @@
+# Overview
+Proposal for exporting KubeVirt Virtual Machine volumes from a Kubernetes cluster
+
+## Motivation
+There are currently multiple ways to import/initialize Virtual Machine volumes.  But there is no easy and efficient method to get volume data out of the cluster.
+
+## Goals
+* Define VirtualMachineExport API
+* Introduce virtctl export command
+
+## Non Goals
+* Implementation details of internal components (servers/proxies/controllers/etc)
+
+## User Stories
+* As a KubeVirt user, I would like to export a Virtual Machine volume for local debugging/analysis.
+* As a KubeVirt user, I would like to export a Virtual Machine volume as part of migrating a Virtual Machine to a different cluster.
+* As a KubeVirt user, I would like to export a Virtual Machine volume to serve as an offsite backup.
+
+## Repos
+* [KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+
+## VirtualMachineExport API
+
+The VirtualMachineExport API allows users to download a Virtual Machine volumes from URLs.
+
+### User Flow
+
+1.  Create a VirtualMachineExport resource
+2.  Wait for the `Ready` condition of the VirtualMachineExport to be `True`
+3.  Download a Virtual Machine volume in the desired format listed under `status.links.volumes`
+
+### Export Formats
+
+| Volume Content Type | Export Format |
+| ------------------- | ------------- |
+| kubevirt            | raw           |
+| kubevirt            | tar.gz        |
+| archive             | dir           |
+| archive             | tar.gz        |
+
+#### raw
+
+The http endpoint supports random access to a raw disk image.
+
+#### tar.gz
+
+The http endpoint is a link to a `tar.gz` streaming download.
+
+#### dir
+
+The http endpoint is a golang `http.FileServer` serving up the root of the filesystem.
+
+### Export a VirtualMachine
```

Comment:
If exporting a `VirtualMachineSnapshot`, first a PVC will be created from each `VolumeSnapshot` it contains.  Then, a pod will be started to serve up the PVC data.  When the `VirtualMachineExport` is deleted, the pod/PVCs will also be deleted.

If exporting a `VirtualMachine` it is expected that the VirtualMachine is stopped or in some sort of maintenance mode.  It should not be possible to export a running VM.

Yamls are already available via the k8s apiserver.

Reply:
> If exporting a VirtualMachine it is expected that the VirtualMachine is stopped or in some sort of maintenance mode. It should not be possible to export a running VM.
>
> Yamls are already available via the k8s apiserver.

I worry that this may be raceful: after the VM is exported and while the data is downloaded, the VM may be started, modifying the data midway. Or the VM definition may be changed. What do you think about taking an implicit snapshot to avoid that?

Reply:
I propose that we start with volume data and can further extend this resource to include manifests and/or a single ova like file containing all vm data and config.

Reply:
It is not obvious to me how this would work. Would you lay out your phased approach in this proposal?